tests/integration_tests/functional/test_logging.py

```diff
@@ -6,69 +6,88 @@
-    """
-    if log_level == "Error":
-        return LOG_LEVELS[0]
-    if log_level == "Warning":
-        return LOG_LEVELS[1]
-    if log_level == "Info":
-        return LOG_LEVELS[2]
-    if log_level == "Debug":
-        return LOG_LEVELS[3]
-    return ""
+    if level == "ERROR":
+        return 0
+    if level == "WARN":
+        return 1
+    if level == "INFO":
+        return 2
+    if level == "DEBUG":
+        return 3
+    if level == "TRACE":
+        return 4
+    raise Exception("Invalid level")
 
 
-def check_log_message_format(log_str, instance_id, level, show_level, show_origin):
+# pylint: disable=anomalous-backslash-in-string
+def check_log_message_format(log_str, instance_id, log_level, show_level, show_origin):
     """Ensure correctness of the logged message.
 
     Parse the string representing the logs and look for the parts
     that should be there.
-    The log line should look lie this:
-         YYYY-MM-DDTHH:MM:SS.NNNNNNNNN [ID:THREAD:LEVEL:FILE:LINE] MESSAGE
-    where LEVEL and FILE:LINE are both optional.
-    e.g. with THREAD NAME as TN
-    `2018-09-09T12:52:00.123456789 [MYID:TN:WARN:/path/to/file.rs:52] warning`
-    """
-    timestamp, tag_and_msg = log_str.split(" ", maxsplit=1)
-    timestamp = timestamp[:-10]
-    strptime(timestamp, "%Y-%m-%dT%H:%M:%S")
-
-    pattern = "\\[(" + instance_id + ")"
-    pattern += ":(.*)"
-    if show_level:
-        pattern += ":(" + "|".join(LOG_LEVELS) + ")"
-    if show_origin:
-        pattern += ":([^:]+/[^:]+):([0-9]+)"
-    pattern += "\\].*"
+    The log line should look like:
+    > {year}-{month}-{day}T{hour}:{minute}:{second}.{nanosecond} [{instance id}:{thread name}:{level}:{file}:{line number}] {message}
+    where `level`, `file` and `line number` are optional e.g.
+    > 2023-07-19T12:10:54.123608814 [fc_api:INFO:src\main.rs:18] yak shaving completed.
+    """
+    split = iter(log_str.split())
+    now = datetime.datetime.now()
+
+    # 2023-07-19T12:10:54.123608814
+    timestamp = next(split)
+    date, time = timestamp.split("T")
+    year, month, day = date.split("-")
+    assert len(month) == 2
+    assert len(day) == 2
+
+    hour, minute, secs = time.split(":")
+    second, nanosecond = secs.split(".")
+    assert len(hour) == 2
+    assert len(minute) == 2
+    assert len(second) == 2
+    assert len(nanosecond) == 9
+
+    # Assert the time in the logs is less than or equal to the current time
+    log_time = datetime.datetime(
+        year=int(year),
+        month=int(month),
+        day=int(day),
+        hour=int(hour),
+        minute=int(minute),
+        second=int(second),
+        microsecond=int(nanosecond) // 1000,
+    )
+    assert log_time <= now
 
-    mo = re.match(pattern, tag_and_msg)
-    assert (
-        mo is not None
-    ), f"Log message ({tag_and_msg}) does not match pattern ({pattern})."
+    # [fc_api:INFO:src\main.rs:18]
+    data = next(split)
+    data_split = iter(data[1:-1].split(":"))
```

Comment:
I'd rather we keep using the regex to extract information, this approach to parsing is less readable IMO

Reply:
See above, it can be done in one line with a regex vs 10+ lines and manually incrementing an iterator.

Reply:
> See above, it can be done in one line with a regex

I think the date can be done with 1 regex match but the rest of the log message requires conditionally composing the string.

I think comparing the original approach for matching the infomation outside the date:

```python
    pattern = "\\[(" + instance_id + ")"
    pattern += ":(.*)"
    if show_level:
        pattern += ":(" + "|".join(LOG_LEVELS) + ")"
    if show_origin:
        pattern += ":([^:]+/[^:]+):([0-9]+)"
    pattern += "\\].*"

    mo = re.match(pattern, tag_and_msg)
    assert (
        mo is not None
    ), f"Log message ({tag_and_msg}) does not match pattern ({pattern})."

    if show_level:
        tag_level = mo.group(3)
        tag_level_no = LOG_LEVELS.index(tag_level)
        configured_level_no = LOG_LEVELS.index(to_formal_log_level(level))
        assert tag_level_no <= configured_level_no
```

With the new one:

```python
    data = next(split)
    data_split = iter(data[1:-1].split(":"))

    log_instance_id = next(data_split)
    assert log_instance_id == instance_id

    # Thread names are not optional.
    _thread_name = next(data_split)

    if show_level:
        level = next(data_split)
        assert level in LOG_LEVELS
        assert LOG_LEVELS[level] <= LOG_LEVELS[log_level.upper()]

    if show_origin:
        _file = next(data_split)
        line = next(data_split)
        assert line.isnumeric()
```

The new approach is simpler.

An example would be that:

```python
if show_origin:
        pattern += ":([^:]+/[^:]+):([0-9]+)"
```
involves a lot more cognitive overhead than
```python
    if show_origin:
        _file = next(data_split)
        line = next(data_split)
        assert line.isnumeric()
```

Reply:
I think if we still want to use the original approach it would make sense to drop the commit that changes the test.